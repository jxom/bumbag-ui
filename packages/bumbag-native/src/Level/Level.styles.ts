import { styled } from '../styled';
import { theme } from '../utils/theme';
import { breakpoint } from '../utils/breakpoint';
import { getAlignmentStyles } from '../utils/getCSSFromStyleObject';

export const StyledLevel = styled.View`
  justify-content: space-between;

  ${getOrientationAttributes}
  ${getResponsiveAttributes}

  ${theme('native.Level', 'styles.base')};
`;

function getOrientationAttributes({ orientation }: any) {
  if (orientation === 'horizontal') {
    return `
      flex-direction: row;
    `;
  }
  if (orientation === 'vertical') {
    return `
      flex-direction: column;
    `;
  }
  return '';
}

function getResponsiveAttributes({ orientation, verticalBelow, ...props }: any) {
  if (orientation === 'horizontal' && verticalBelow) {
    return breakpoint('width', {
      [`max-${verticalBelow}`]: `
        justify-content: flex-start;
        flex-direction: column;

        ${props.alignY ? getAlignmentStyles({ attribute: 'alignY', value: props.alignY, props }) : ''}
        ${props.alignX ? getAlignmentStyles({ attribute: 'alignX', value: props.alignX, props }) : ''}
      `,
    })(props);
  }
  return '';
}
