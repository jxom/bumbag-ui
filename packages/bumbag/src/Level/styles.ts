import { css, cssClass } from '../styled';
import { breakpoint, space, theme } from '../utils';
import { getAlignmentAttributes } from '../Box/styles';

const verticalBreakpoints = {
  tablet: 'mobile',
  desktop: 'tablet',
  widescreen: 'desktop',
  fullHD: 'widescreen',
};

export const Level = (styleProps) => cssClass`
  &&& {
    ${breakpoint(styleProps.verticalBelow ? `max-${verticalBreakpoints[styleProps.verticalBelow]}` : null, css`
    flex-direction: column;

    & > *:not(:last-child) {
      margin-bottom: ${space(styleProps.spacing)(styleProps)}rem;
    }

    ${getAlignmentAttributes(styleProps)}
    `, { else: css`
    justify-content: space-between;
    ` })(styleProps)};
  }


  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  }
`;
