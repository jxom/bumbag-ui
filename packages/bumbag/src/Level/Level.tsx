import { Box as ReakitBox } from 'reakit';

import { Breakpoint } from '../types';
import { useClassName, createComponent, createElement, createHook } from '../utils';
import { Flex, FlexProps } from '../Flex';

import * as styles from './styles';

export type LocalLevelProps = {
  /** Sets the orientation of the level. */
  orientation?: 'vertical' | 'horizontal';
  /** Sets the spacing of the level when it snaps to a vertical orientation. */
  spacing?: string;
  /** Sets the breakpoint at which the level should become vertical. */
  verticalBelow?: Breakpoint;
};
export type LevelProps = FlexProps & LocalLevelProps;

const useProps = createHook<LevelProps>(
  (props, { themeKey, themeKeyOverride }) => {
    const flexProps = Flex.useProps(props);

    const className = useClassName({
      style: styles.Level,
      styleProps: props,
      themeKey,
      themeKeyOverride,
      prevClassName: flexProps.className,
    });

    return { ...flexProps, className };
  },
  {
    defaultProps: { alignX: 'center', orientation: 'horizontal', spacing: 'major-2', verticalBelow: 'tablet' },
    themeKey: 'Level',
  }
);

export const Level = createComponent<LevelProps>(
  (props) => {
    const textProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: textProps });
  },
  {
    attach: {
      useProps,
    },
    themeKey: 'Level',
  }
);
