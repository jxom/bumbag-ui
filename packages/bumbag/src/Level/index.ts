import * as styles from './styles';

export * from './Level';
export { styles as levelStyles };
