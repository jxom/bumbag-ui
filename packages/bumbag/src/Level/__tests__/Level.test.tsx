import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Level } from '../Level';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Level>Hello world</Level>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<Level color="primary">Hello world</Level>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('variants', () => {
  it('styles.base should render correctly', () => {
    const { container } = render(<Level variant="test">hello world</Level>, {
      theme: { Level: { variants: { test: { styles: { base: { backgroundColor: 'red' } } } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('overrides', () => {
  it('styles.base should render correctly', () => {
    const { container } = render(
      <Level overrides={{ Level: { styles: { base: { backgroundColor: 'red' } } } }}>hello world</Level>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('styles.base should render correctly', () => {
    const { container } = render(<Level>hello world</Level>, {
      theme: { Level: { styles: { base: { backgroundColor: 'red' } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(<Level>hello world</Level>, {
      theme: { Level: { defaultProps: { className: 'test', color: 'primary' } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
