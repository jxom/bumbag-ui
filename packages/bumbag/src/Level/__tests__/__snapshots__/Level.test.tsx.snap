// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultProps should render correctly for className 1`] = `
@media screen and (max-width: 767px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .emotion-0.emotion-0.emotion-0>*:not(:last-child) {
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1.emotion-1 {
  -webkit-flex-direction: unset;
  -ms-flex-direction: unset;
  flex-direction: unset;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: unset;
  -webkit-box-align: unset;
  -ms-flex-align: unset;
  align-items: unset;
}

.emotion-2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  color: var(--bb-palette-primary, #000000)!important;
}

<div
  class="emotion-0 emotion-1 emotion-2 emotion-3 test bb-Box bb-Flex bb-Level"
>
  hello world
</div>
`;

exports[`overrides styles.base should render correctly 1`] = `
@media screen and (max-width: 767px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .emotion-0.emotion-0.emotion-0>*:not(:last-child) {
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-0 {
  background-color: var(--bb-palette-red, #ff0000);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1.emotion-1 {
  -webkit-flex-direction: unset;
  -ms-flex-direction: unset;
  flex-direction: unset;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: unset;
  -webkit-box-align: unset;
  -ms-flex-align: unset;
  align-items: unset;
}

.emotion-2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="emotion-0 emotion-1 emotion-2 emotion-3 bb-Box bb-Flex bb-Level"
>
  hello world
</div>
`;

exports[`props should render correctly 1`] = `
@media screen and (max-width: 767px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .emotion-0.emotion-0.emotion-0>*:not(:last-child) {
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1.emotion-1 {
  -webkit-flex-direction: unset;
  -ms-flex-direction: unset;
  flex-direction: unset;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: unset;
  -webkit-box-align: unset;
  -ms-flex-align: unset;
  align-items: unset;
}

.emotion-2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="emotion-0 emotion-1 emotion-2 emotion-3 bb-Box bb-Flex bb-Level"
>
  Hello world
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
@media screen and (max-width: 767px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .emotion-0.emotion-0.emotion-0>*:not(:last-child) {
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1.emotion-1 {
  -webkit-flex-direction: unset;
  -ms-flex-direction: unset;
  flex-direction: unset;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: unset;
  -webkit-box-align: unset;
  -ms-flex-align: unset;
  align-items: unset;
}

.emotion-2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  color: var(--bb-palette-primary, #000000)!important;
}

<div
  class="emotion-0 emotion-1 emotion-2 emotion-3 bb-Box bb-Flex bb-Level"
>
  Hello world
</div>
`;

exports[`theming styles.base should render correctly 1`] = `
@media screen and (max-width: 767px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .emotion-0.emotion-0.emotion-0>*:not(:last-child) {
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-0 {
  background-color: var(--bb-palette-red, #ff0000);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1.emotion-1 {
  -webkit-flex-direction: unset;
  -ms-flex-direction: unset;
  flex-direction: unset;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: unset;
  -webkit-box-align: unset;
  -ms-flex-align: unset;
  align-items: unset;
}

.emotion-2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="emotion-0 emotion-1 emotion-2 emotion-3 bb-Box bb-Flex bb-Level"
>
  hello world
</div>
`;

exports[`variants styles.base should render correctly 1`] = `
@media screen and (max-width: 767px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .emotion-0.emotion-0.emotion-0>*:not(:last-child) {
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0.emotion-0.emotion-0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-0 {
  background-color: var(--bb-palette-red, #ff0000);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1.emotion-1 {
  -webkit-flex-direction: unset;
  -ms-flex-direction: unset;
  flex-direction: unset;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: unset;
  -webkit-box-align: unset;
  -ms-flex-align: unset;
  align-items: unset;
}

.emotion-2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-2:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="emotion-0 emotion-1 emotion-2 emotion-3 bb-Box bb-Flex bb-Level"
>
  hello world
</div>
`;
